c++/src/kj/compat/http.c++

```diff
@@ -3951,6 +3952,149 @@ kj::Own<HttpClient> newHttpClient(kj::Timer& timer, HttpHeaderTable& responseHea
 
 namespace {
 
+class ConcurrencyLimitingHttpClient final: public HttpClient {
+public:
+  ConcurrencyLimitingHttpClient(kj::HttpClient& inner, uint maxConcurrentRequests)
+      : inner(inner),
+        maxConcurrentRequests(maxConcurrentRequests) {}
+
+  Request request(HttpMethod method, kj::StringPtr url, const HttpHeaders& headers,
+                  kj::Maybe<uint64_t> expectedBodySize = nullptr) override {
+    if (concurrentRequests < maxConcurrentRequests) {
+      auto counter = kj::heap<ConnectionCounter>(*this);
+      auto request = inner.request(method, url, headers, expectedBodySize);
+      auto promise = attachCounter(kj::mv(request.response), kj::mv(counter));
+      return { kj::mv(request.body), kj::mv(promise) };
+    }
+
+    auto paf = kj::newPromiseAndFulfiller<kj::Own<ConnectionCounter>>();
+    auto urlCopy = kj::str(url);
+    auto headersCopy = headers.clone();
+    // TODO(now): Use PromiseOutputStream instead
+    auto pipe = newOneWayPipe(expectedBodySize);
+
+    auto promise = paf.promise
+        .then([this,
+               method,
+               urlCopy = kj::mv(urlCopy),
+               headersCopy = kj::mv(headersCopy),
+               expectedBodySize,
+               pipeIn = kj::mv(pipe.in)](kj::Own<ConnectionCounter>&& counter) mutable {
+          auto request = inner.request(method, urlCopy, headersCopy, expectedBodySize);
+          return pipeIn->pumpTo(*request.body).ignoreResult().attach(kj::mv(request.body))
+              .then([response = kj::mv(request.response),
+                     counter = kj::mv(counter)]() mutable {
+                return attachCounter(kj::mv(response), kj::mv(counter));
+              });
+        });
+
+    pendingRequests.push(kj::mv(paf.fulfiller));
+    return { kj::mv(pipe.out), kj::mv(promise) };
+  }
+
+  kj::Promise<WebSocketResponse> openWebSocket(
+      kj::StringPtr url, const kj::HttpHeaders& headers) override {
+    if (concurrentRequests < maxConcurrentRequests) {
+      auto counter = kj::heap<ConnectionCounter>(*this);
+      return attachCounter(inner.openWebSocket(url, headers), kj::mv(counter));
+    }
+
+    auto paf = kj::newPromiseAndFulfiller<kj::Own<ConnectionCounter>>();
+    auto urlCopy = kj::str(url);
+    auto headersCopy = headers.clone();
+
+    auto promise = paf.promise
+        .then([this,
+               urlCopy = kj::mv(urlCopy),
+               headersCopy = kj::mv(headersCopy)](kj::Own<ConnectionCounter>&& counter) mutable {
+          return attachCounter(inner.openWebSocket(urlCopy, headersCopy), kj::mv(counter));
+        });
+
+    pendingRequests.push(kj::mv(paf.fulfiller));
+    return kj::mv(promise);
+  }
+
+private:
+  struct ConnectionCounter;
+
+  kj::HttpClient& inner;
+  uint maxConcurrentRequests;
+  uint concurrentRequests = 0;
+
+  std::queue<kj::Own<kj::PromiseFulfiller<kj::Own<ConnectionCounter>>>> pendingRequests;
+  // TODO(someday): want maximum cap on queue size?
+
+  struct ConnectionCounter final {
```

Comment:
I recommend adding:

    KJ_DISALLOW_COPY(ConnectionCounter);

just to be safe.

FWIW, another way to design this would be to make ConnectionCounter a movable type, with `parent` being a pointer that becomes null when moved away. Then it wouldn't need to be allocated on the heap. However, the code would be somewhat more complicated and the performance gain probably insignificant... so meh.